### PR TITLE
Update text to one-key-for-all-resources

### PR DESCRIPTION
**Pull request by sttts**

### CODE REVIEW

File: apiserver/encrypting-data-at-datastore-layer.md [hunk 1]
```diff
@@ -126,100 +126,144 @@ User enables encryption by setting `apiserver.spec.encryption.type` to `aescbc`.
...
+
+- **gets** all API server pods by the matching `encryption.operator.openshift.io/component` label to
+  - to check whether they converged 
+  - and to know to which revision.
+- **gets** the `<operand-target-namespace>/encryption-config-<revision>` secret.
+- **gets** the keys in `openshift-config-managed` by matching `encryption.operator.openshift.io/component` label.
+
+and then following:
+
+- if to-be-encrypted GRs are missing in the old config, return desired config with (a) updated read-keys for all existing resources (b) added to-be-encrypted resourcdes with all the read-keys and `identity` for the write key.
```
Comment (enj) [inline apiserver/encrypting-data-at-datastore-layer.md]: `resourcdes` typo

File: apiserver/encrypting-data-at-datastore-layer.md [hunk 2]
```diff
@@ -126,100 +126,144 @@ User enables encryption by setting `apiserver.spec.encryption.type` to `aescbc`.
...
+- if configured read-keys do not match the **created** keys in the cluster, return desired config with updated read-keys.
+- if a write-key does not match the latest **observed read-key**, return desired config with this latest **observed read-key** as write-key.
+- if the write-key is marked as **migrated** for all **encrypted** resources, return desired config with all other read-keys removed.
```
Comment (enj) [inline apiserver/encrypting-data-at-datastore-layer.md]: is this encrypted and to-be-encrypted?

File: apiserver/encrypting-data-at-datastore-layer.md [hunk 3]
```diff
@@ -126,100 +126,144 @@ User enables encryption by setting `apiserver.spec.encryption.type` to `aescbc`.
...
+
+##### encryptionKeyController
+
+The `encryptionKeyController` implements the `->1` transition. It
+
+- **watches** pods and secrets in `<operand-target-namespace>` (and is triggered by changes)
+- **computes** a new, desired encryption config from `encryption-config-<revision>` and the existing keys in `openshift-config-managed`. 
+- **derives** from the desired encryption config whether a new key is needed. It then creates it.
```
Comment (enj) [inline apiserver/encrypting-data-at-datastore-layer.md]: `whether a new key is needed` should we expand on this?

File: apiserver/encrypting-data-at-datastore-layer.md [hunk 4]
```diff
@@ -126,100 +126,144 @@ User enables encryption by setting `apiserver.spec.encryption.type` to `aescbc`.
...
+- **updates** the `encryption.operator.openshift.io/migrated-timestamp` and  `encryption.operator.openshift.io/migrated-resources` annotations on the write-key secrets used for migration in the previous step.
 
-1. The mapping of secret to actual encryption key
-1. The mapping of actual encryption key to secret
-1. The read, write and migrated state of each secret
-1. The last secret to be migrated
+##### encryptionPruneController
 
-Translation of each `GroupResource`'s state to individual `ResourceConfiguration`s within the complete `EncryptionConfiguration`:
+`encryptionPruneController` prevents an unbounded growth of old encryption keys.
```
Comment (enj) [inline apiserver/encrypting-data-at-datastore-layer.md]: it actually just nukes everything but the write key now

File: apiserver/encrypting-data-at-datastore-layer.md [hunk 5]
```diff
@@ -126,100 +126,144 @@ User enables encryption by setting `apiserver.spec.encryption.type` to `aescbc`.
...
-1. A controller that deletes old encryption keys from `openshift-config-managed` as needed (to prevent old keys from building up).  It first removes the `encryption.apiserver.operator.openshift.io/deletion-protection` finalizer and then removes the secret itself.  Thus if a user accidentally runs a command like `oc delete secret --all -n openshift-config-managed`, it will not result in the cluster becoming permanently inaccessible.
-1. A controller that runs storage migration for new write keys (this is what actually causes data to be encrypted under a given write key).  This migration only runs when all API servers are on the same revision to prevent writing of data to etcd with a key that is not observed by all servers.  After the migration is complete, the migration timestamp annotation is set to the current time.
+The `encryptionStateController` controller implements transitions `1->2`, `2-3`, and `4-5` (does it????). It
```
Comment (enj) [inline apiserver/encrypting-data-at-datastore-layer.md]: on the `4->5`, it drops the key from the encryption config, which causes the prune controller to delete it.  so it starts the transition and the prune controller completes it

File: apiserver/encrypting-data-at-datastore-layer.md [hunk 6]
```diff
@@ -126,100 +126,144 @@ User enables encryption by setting `apiserver.spec.encryption.type` to `aescbc`.
...
 
-kube-apiserver:
+#### Controllers
 
-1. `secrets`
-1. `configmaps`
+The transitions are implemented by a series of controllers which each perform a single, _simple_ task.
 
-openshift-apiserver:
+##### Shared code: computing a new desired configuration
```
Comment (enj) [inline apiserver/encrypting-data-at-datastore-layer.md]: risk section needs to point out that this handles upgrade with new resources correctly but not downgrade

File: apiserver/encrypting-data-at-datastore-layer.md [hunk 7]
```diff
@@ -126,100 +126,144 @@ User enables encryption by setting `apiserver.spec.encryption.type` to `aescbc`.
...
-with the mandatory and authoritative labels:
+#### State
 
-- `encryption.apiserver.operator.openshift.io/component`
-- `encryption.apiserver.operator.openshift.io/group`
-- `encryption.apiserver.operator.openshift.io/resource`
+The distributed state machine is using the following data:
 
-and annotations to track the current state of the secret (explained further below):
+1. keys stored in `openshift-config-managed/<component>-encryption-<unsigned integer #>` secrets, and found via the label `encryption.operator.openshift.io/component` for the respective component. They are named as `key-<unsigned integer #>`.
```
[the remark this answers is not in the document]
Comment (sttts) [reply apiserver/encrypting-data-at-datastore-layer.md]: where? Still see the upper on in the code.

File: apiserver/encrypting-data-at-datastore-layer.md [hunk 8]
```diff
@@ -126,100 +126,144 @@ User enables encryption by setting `apiserver.spec.encryption.type` to `aescbc`.
...
-1. **created** - the key secret is created
-1. **read-key** - all API servers have observed and configured the key as a read key
-1. **write-key** - all API servers have observed and configured the key as a write key
-1. **migrated** - since the key has been write-key, all key-values in etcd have been rewritten (=migrated)
-1. **deleted** - all API servers have removed the key as read key
+The encryption configuration maintains the following invariants
 
-They go through 1-4 via these transitions:
+- the read-keys for all **encrypted** GRs are the same.
+- there is at most one non-identity write-key for all **encrypted** GRs (e.g. a new resource starts with `identity`, all others have a `aescbc` key already).
```
[the remark this answers is not in the document]
Comment (sttts) [reply apiserver/encrypting-data-at-datastore-layer.md]: logically correct as it is. If encryption is not enabled, there is no encrypted GR.

File: apiserver/encrypting-data-at-datastore-layer.md [hunk 9]
```diff
@@ -126,100 +126,144 @@ User enables encryption by setting `apiserver.spec.encryption.type` to `aescbc`.
...
 
-After another key has been migrated to, i.e. reaches 4, an old key can be decommissioned:
-- 4 -> 5: the old key secret in the cluster is deleted and asynchronously all API servers are updated on the master nodes (no need to wait for that to finish as the read-key is not used anymore after 4).
+The life-cycle of a key implemented by the controllers is: 1, 2, 3, 4, 5. 
 
-##### Resources
+A transition only takes place when all API servers have converged to the same revision, are running, and hence use the same encryption configuration.
 
-This list of resources is not configurable.  The following resources are encrypted:
+- ->1: a new key secret is created if
```
[the remark this answers is not in the document]
Comment (sttts) [reply apiserver/encrypting-data-at-datastore-layer.md]: In the time-based case I understand this. Why do we wait for migration for the other cases?

File: apiserver/encrypting-data-at-datastore-layer.md [hunk 10]
```diff
@@ -126,100 +126,144 @@ User enables encryption by setting `apiserver.spec.encryption.type` to `aescbc`.
...
 
-This list of resources is not configurable.  The following resources are encrypted:
+- ->1: a new key secret is created if 
+  - encryption is enabled via the API or
+  - a new **to-be-encrypted** resource shows up or
+  - the `EncryptionType` in the API does not match with the newest existing key or
+  - based on time (once a **week** is the proposed rotation interval).
+- 1->2: when a new **created** key is found.
+- 2->3: when all API servers have **observed the read-key for resource GR**.
+- 3->4: when all GR instances have been rewritten in etcd and the key is marked as **migrated** for that GR by the migration mechanism.
```
[the remark this answers is not in the document]
Comment (sttts) [reply apiserver/encrypting-data-at-datastore-layer.md]: Migration is still per resource. A resource that is new on upgrade, is not migrated yet.

File: apiserver/encrypting-data-at-datastore-layer.md (hunk 2 above)
Comment (sttts) [reply apiserver/encrypting-data-at-datastore-layer.md]: both

File: apiserver/encrypting-data-at-datastore-layer.md (hunk 3 above)
Comment (sttts) [reply apiserver/encrypting-data-at-datastore-layer.md]: references `->1`.

File: apiserver/encrypting-data-at-datastore-layer.md (hunk 4 above)
Comment (sttts) [reply apiserver/encrypting-data-at-datastore-layer.md]: where are old read keys pruned from the config?

File: apiserver/encrypting-data-at-datastore-layer.md (hunk 5 above)
Comment (sttts) [reply apiserver/encrypting-data-at-datastore-layer.md]: where is that code which drops old read keys?

File: apiserver/encrypting-data-at-datastore-layer.md (hunk 6 above)
Comment (sttts) [reply apiserver/encrypting-data-at-datastore-layer.md]: please formulate something.

My thoughts: we have to store the set of resources somewhere. We actually do in the encryption config already. A resource that is encrytped (i.e. not to-be-encrypted), must always stay in that set.

File: apiserver/encrypting-data-at-datastore-layer.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (sttts) [reply apiserver/encrypting-data-at-datastore-layer.md]: encryption.apiserver.operator.openshift.io/component

**Pull request merged**